### ISSUE TITLE
Webapp cluster performance issue

### ISSUE THREAD

**Issue by Kenterfie** [graphite-project/graphite-web#1690]
**Configuration**
OS: Ubuntu 14.04
Webserver: nginx
WSGI: uwsgi
Graphite: graphite (latest git)

2 Graphite storage server + 1 Graphite Webapp Render Host (only json output)

**Issue**
We save our data on two different graphite storage hosts over statsite + carbon-c-relay . Everything is working well, but one problem we notice is, that every request needs very long on webapp render host.

Rendering performance
Small request needs >1000ms on webapp render host, but the real rendering on the storage host needs only ~15ms. What does graphite webapp the other 980ms on the render host?

When we call the same render request over wget from the render host it needs only 180ms not 1000ms. Someone else notice render performance issues like this? I havnt found any reason in graphite currently, what cause it.

**Comment by deniszh** [comment]
When you say 'latest git' is it really latest, i.e. 2aee2bc ? And it includes 2cc2a2a ?

**Comment by Kenterfie** [comment]
I have everything updated 3 hours ago, so yes. The latest commit is 2aee2bc, but the problem was with 0.9.15 the same.

**Comment by deniszh** [comment]
>Small request needs >1000ms on webapp render host
>When we call the same render request over wget from the render host it needs only 180ms not 1000ms. 

Sorry, looks really confusing, do not really understand which request takes what. Could you maybe give exact queries?

**Comment by deniszh** [comment]
>I have everything updated 3 hours ago, so yes. The latest commit is 2aee2bc, but the problem was with 0.9.15 the same.

Ah, then maybe it's something else.  Did you have `REMOTE_STORE_MERGE_RESULTS` enabled on 0.9.15 ?

**Comment by deniszh** [comment]
Another (probably stupid) question - are you sure that you really running on master code now, and not on still 0.9.15 ?

**Comment by Kenterfie** [comment]
In my case not necessary. I don't use replication between the two graphite hosts. Both hosts holding different data. I use the render node to have one api endpoint.

And yes, im very sure i use the latest version. I have compared my webapp folder with the latest git master.

**Comment by Kenterfie** [comment]
On render host
Thu Sep 08 17:27:57 2016 :: Retrieval of sortByMaxima(summarize(..)) took 0.900179
Thu Sep 08 17:27:57 2016 :: Retrieval of sortByMaxima(summarize(..)) took 0.893999
Thu Sep 08 17:27:58 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.089880
Thu Sep 08 17:27:58 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.084184
Thu Sep 08 17:27:59 2016 :: Retrieval of sortByMaxima(summarize(..)) took 0.905862
Thu Sep 08 17:27:59 2016 :: Retrieval of sortByMaxima(summarize(..)) took 0.900468
Thu Sep 08 17:28:00 2016 :: Retrieval of sortByMaxima(summarize(..)) took 0.901901
Thu Sep 08 17:28:00 2016 :: Retrieval of sortByMaxima(summarize(..)) took 0.900201
Thu Sep 08 17:28:01 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.088676
Thu Sep 08 17:28:01 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.084956
Thu Sep 08 17:28:03 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.081588
Thu Sep 08 17:28:03 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.357989
Thu Sep 08 17:28:04 2016 :: Retrieval of sortByMaxima(summarize(..)) took 0.908793
Thu Sep 08 17:28:04 2016 :: Retrieval of sortByMaxima(summarize(..)) took 0.897626
Thu Sep 08 17:28:05 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.374054
Thu Sep 08 17:28:06 2016 :: Retrieval of sortByMaxima(summarize(..)) took 2.076580
Thu Sep 08 17:28:06 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.084878
Thu Sep 08 17:28:07 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.082674
Thu Sep 08 17:28:07 2016 :: Retrieval of sortByMaxima(summarize(..)) took 0.910757
Thu Sep 08 17:28:08 2016 :: Retrieval of sortByMaxima(summarize(..)) took 0.899178
Thu Sep 08 17:28:09 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.102875
Thu Sep 08 17:28:09 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.083390
Thu Sep 08 17:28:10 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.099460
Thu Sep 08 17:28:10 2016 :: Retrieval of sortByMaxima(summarize(..)) took 1.080182

And on the storage node the forwarded request from render host
Thu Sep 08 17:30:07 2016 :: Retrieval of .. took 0.013766
Thu Sep 08 17:30:08 2016 :: Retrieval of .. took 0.015371
Thu Sep 08 17:30:08 2016 :: Retrieval of .. took 0.015578
Thu Sep 08 17:30:11 2016 :: Retrieval of .. took 0.015855
Thu Sep 08 17:30:11 2016 :: Retrieval of .. took 0.008226
Thu Sep 08 17:30:12 2016 :: Retrieval of .. took 0.014218
Thu Sep 08 17:30:12 2016 :: Retrieval of .. took 0.011763
Thu Sep 08 17:30:13 2016 :: Retrieval of .. took 0.013829
Thu Sep 08 17:30:13 2016 :: Retrieval of .. took 0.017468
Thu Sep 08 17:30:14 2016 :: Retrieval of .. took 0.014424
Thu Sep 08 17:30:14 2016 :: Retrieval of .. took 0.012316
Thu Sep 08 17:30:15 2016 :: Retrieval of .. took 0.010825
Thu Sep 08 17:30:15 2016 :: Retrieval of .. took 0.015276
Thu Sep 08 17:30:17 2016 :: Retrieval of .. took 0.014506
Thu Sep 08 17:30:18 2016 :: Retrieval of .. took 0.013076
Thu Sep 08 17:30:18 2016 :: Retrieval of .. took 0.010940
Thu Sep 08 17:30:19 2016 :: Retrieval of .. took 0.010102
Thu Sep 08 17:30:21 2016 :: Retrieval of .. took 0.009397
Thu Sep 08 17:30:21 2016 :: Retrieval of .. took 0.010097
Thu Sep 08 17:30:22 2016 :: Retrieval of .. took 0.012833
Thu Sep 08 17:30:22 2016 :: Retrieval of .. took 0.014324

**Comment by Kenterfie** [comment]
other ideas what cause the hugh delay between data gathering and data delivery?

**Comment by Kenterfie** [comment]
Here are a exact query
Thu Sep 08 20:15:33 2016 :: Retrieval of sortByMaxima(summarize(statsite.machines.{xy462}.gz1029059_1.{memoryUsage,cpuUsage},'1d','avg')) took 1.083491

**Comment by Kenterfie** [comment]
Ok i think i found the issue

Thu Sep 08 20:28:42 2016 :: Retrieval of summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg') took 0.194305
Thu Sep 08 20:28:53 2016 :: Retrieval of summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg') took 0.010299
Thu Sep 08 20:28:53 2016 :: Retrieval of summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg') took 0.010212
Thu Sep 08 20:28:53 2016 :: Retrieval of summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg') took 0.016453
Thu Sep 08 20:28:53 2016 :: Retrieval of summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg') took 0.192053
Thu Sep 08 20:28:53 2016 :: Retrieval of summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg') took 0.011103
Thu Sep 08 20:28:53 2016 :: Retrieval of summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg') took 0.016286
Thu Sep 08 20:30:10 2016 :: Retrieval of sortByMaxima(summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg')) took 0.903806
Thu Sep 08 20:30:12 2016 :: Retrieval of sortByMaxima(summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg')) took 1.907195
Thu Sep 08 20:30:13 2016 :: Retrieval of sortByMaxima(summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg')) took 1.086445
Thu Sep 08 20:30:15 2016 :: Retrieval of sortByMaxima(summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg')) took 0.903029
Thu Sep 08 20:30:16 2016 :: Retrieval of sortByMaxima(summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg')) took 1.091595
Thu Sep 08 20:30:17 2016 :: Retrieval of sortByMaxima(summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg')) took 0.924130
Thu Sep 08 20:30:18 2016 :: Retrieval of sortByMaxima(summarize(statsite.nodeXY.{memoryUsage,cpuUsage},'1d','avg')) took 0.902272

sortByMaxima is 10 times slower, without a reason in my opinion. the returned data is very small. should not take 1 second

**Comment by Kenterfie** [comment]
Ok, the above code was distorted by caching. The problem still consist

**Comment by deniszh** [comment]
Graphite clustering is quite complicated. I never tried to measure timings and compare them between frontend and backend. 0.9.15 works quite good, I didn't try it on master though, but other people using clustering on master too.
Is that regression visible only on summarize() or globs or on any requests?

**Comment by Kenterfie** [comment]
I had tried to use 0.9.15, but it has to many bugs which are only fixed in later commits and dont like to merge every necessary commit himself to get a "stable" version. The regression is visible on all kinds of requests. I have tried now different ones and all very slow compared to http request to storage host directly. I have written now a small script to bypass this behaviour in this specific subject, but i hope later a performance improved version is released and someone found the reason for it. When i had more time i would do it himself.

**Comment by deniszh** [comment]
@Kenterfie - did you try 0.9.x branch instead of 0.9.15?
Did I understand you right - cluster on current master is still much slower than on 0.9.15?

>I dont know, why no stable version is released since month to fix the most annoying bugs in 0.9.15
IIRC that 0.9.15 should be last 0.9.x release. If we had proper semver semantic that 0.9.16 should be bugfix release - but we're not have it.
Maybe releasing 0.9.16 is a good idea. What do you think, @obfuscurity ?

**Comment by deniszh** [comment]
Just test that on 0.9.15 (with proper waiting for 10 minutes between requests to mitigate the cache), for `format=json` and quite complicated target with 3 globs -  `aliasByNode(nonNegativeDerivative(groupByNode(node*.*.blah*.value, 1, 'sumSeries')), 0)"`
Results - 
Backend 1 - 1.5 sec
Backend 2 - 1.6 sec
Frontend (with just 2 backends above) - 15.3 sec 
As I said, clustering is hard. If you can avoid that - do that.

**Comment by Cka3o4Huk** [comment]
I've faced issue with sortByMaxima() too, server went down due to lack of RAM.

**Comment by redbaron** [comment]
How do you run graphite? You have to make sure, that pool of workers graphite talks too is different from pool of workers browsers/end clients talk to. Otherwise extensive wait times and deadlocks possible

**Comment by deniszh** [comment]
@redbaron: it's a nice thing to have dedicated caches for graphite-web, but that's not mandatory. You can run cluster w/o any problems up to some extent on shared caches.

**Comment by redbaron** [comment]
unfortunately it is mandatory. Simple example:

2 nodes 1 worker thread each, each is configured to talk to same port when making graphite -> graphite requests.

2 browsers requests come in, each Graphite-web tries to contact another graphite and these requests are waiting while thread becomes free and ready to serve request. Everything just deadlocks until timeouts start to kick in and when they do both requests return error.

It doesn't really matter how many threads you run, as long as there is enough activity from browsers/clients and given that each request there causes 2*(N-1) requests to other graphite servers (first to find metric and next to fetch it), it's not long until deadlocks like described above start happening.

**Comment by deniszh** [comment]
That's really nice thought experiment, but I have 0 issues on my graphite clusters, at least with load 0.5 - 1.5 M metrics/min write and 2000-5000 K metrics/min read.
So, no, that's not mandatory. :)

**Comment by redbaron** [comment]
we had deadlocks all over, whole cluster was coming down to a halt until I separated pools

**Comment by deniszh** [comment]
So, that's my point - YMMV. Nice to try separate pools in case of deadlocks, but 99% of users never hit that problem.
Also, IIRC you have quite a big load, around 6M/min? even hitting cache problems in go-carbon - maybe that's case.

**Comment by deniszh** [comment]
Wondering also how you can hit deadlocks on twisted-based daemon... Maybe because of some race conditions or other bugs in carbon or twisted.

**Comment by redbaron** [comment]
My setup:
8  nodes,  12M metrics/minute received with replication factor of 2, so 24M /minute persisted, graphite is run with apache + mod_wsgi, but I can't see how twisted will be different here.

there are not that many requests coming from browsers: 1200-1500 / minute, but graphs are quite big, with json response sizes being as high as 500KB (10% are >225KB).

I  configure apache with 2 WSGI process groups on ports 8000 and 8010, both pointing to the same graphite install with same with `CLUSTER_SERVERS` in local_settings.py pointing to port 8010.

This way workers on port 8000 always waiting for workers on ports 8010 and never other way aroud. It eliminated lockups and improved responsiveness (although I didn't measure response time effect, so can't prove it with numbers)

**Comment by obfuscurity** [comment]
@redbaron Can you provide a diagram of your setup?

**Comment by deniszh** [comment]
>8 nodes, 12M metrics/minute received with replication factor of 2, so 24M /minute persisted

24M per 8 nodes,  i.e. 3M per node, right?

>there are not that many requests coming from browsers: 1200-1500 / minute, but graphs are quite big, with json response sizes being as high as 500KB (10% are >225KB).

Do you have maxDataPoints implemented btw? Just checked - even on 6-month graph I see 100-200K responses max. 

> I configure apache with 2 WSGI process groups on ports 8000 and 8010, both pointing to the same graphite install with same with CLUSTER_SERVERS in local_settings.py pointing to port 8010.

Ah, you mean separate pools of graphite-web in WSGI config for a client and intra-cluster setup? Got it. I thought separate carbon daemon for carbonlink queries and writes.
I never have this problem on my load, nor on Apache + WSGI nor on Nginx + Gunicorn - but I think you can simply increase a number of workers on port 8000.

**Comment by reyjrar** [comment]
Can confirm this with the upgrade to master, I pulled yesterday to ed04b23.  The only difference being 0.9.15 to master.  Render times are noticeably through the roof.  I didn't change the config and had CLUSTER_SERVERS enabled in the past to merge results.  I'll have a graph of the performance degradation soon, but it's with all queries, not just sortByMaxima.  I can produce the results with a *single* metric with no transforms.

**Comment by obfuscurity** [comment]
I don't know if anyone noticed this the first time, but @cbowman0's https://github.com/graphite-project/graphite-web/issues/1063#issuecomment-240763266 appears to demonstrate a 2s latency in cluster requests in the _best case_ and in one case (`datastore_3`), it looks like 40s.

@bmhatfield does that look correct to you?

**Comment by cbowman0** [comment]
That cluster wasn't in the best of states when I ran those tests.  It's my disaster recovery/standby cluster that needed/needs reconfiguration to be viable.  It has full input load for carbon, but essentially no read load so I could do isolated traffic capture of graphite-web requests.

More than a couple carbon caches on at least datastore_3 were at MAX_CACHE_SIZE and that seems to cause the carbon link lookups to timeout frequently on my carbon build (old megacarbon fork).  Since local fetches to whisper files are done serially, and thus carbonlink lookups for reach one, that request was expected to be extra long.  Essentially, those numbers are probably closer to worse case than best.

I've been meaning to upgrade graphite-web to after the merge result set changes were merged. I will attempt to get some timings of some queries to see if I can see any difference in performance in my setup.

**Comment by obfuscurity** [comment]
Ok that's fair. I think it's still fair to acknowledge we have some performance issues afa parallel cluster requests. I expect to be fairly occupied over the next couple days but I'll try to setup a test cluster scenario this weekend.

**Comment by deniszh** [comment]
>Render times are noticeably through the roof. I didn't change the config and had CLUSTER_SERVERS enabled in the past to merge results.

@reyjrar, could you please test does REMOTE_STORE_MERGE_RESULTS = False make things better for master or not?
I'm trying to find is that a mail culprit now or it's because of lack of parallel requests.
Have plans to test rendering on master too on one of my clusters, but not sure if will have time this week...

**Comment by reyjrar** [comment]
@deniszh REMOTE_STORE_MERGE_RESULTS = False had zero effect

Here's render times with CLUSTER_SERVERS commented out:

> Thu Sep 15 14:22:48 2016 :: Total rendering time 0.161674 seconds
> Thu Sep 15 14:22:50 2016 :: Total rendering time 0.169428 seconds

Adding it back in with REMOTE_STORE_MERGE_RESULTS = False

> Thu Sep 15 14:23:39 2016 :: Total rendering time 4.806623 seconds
> Thu Sep 15 14:23:48 2016 :: Total rendering time 4.845406 seconds

**Comment by reyjrar** [comment]
FWIW, I just git reset --hard 6e6c77abd2fad247d97b6a0adce8b595d2f7daba to rule out @redbaron 's patch.  The bug exists prior to that commit.

**Comment by reyjrar** [comment]
My setup: 8 carbon-cache instances per node, 4 nodes.  carbon-c-relays in front doing aggregation and sending to all 8 carbon-caches.  Graphite-web is configured to talk to the same 8 backend carbon-caches.

Re-reading this thread, it sounds like I need to spin up extra cache instances for graphite-web?

**Comment by deniszh** [comment]
@reyjrar - could you please test this PR https://github.com/graphite-project/graphite-web/pull/1699 ?
I tried to tag you there but seems misread your github handle...

**Comment by reyjrar** [comment]
Sure thing, I'll test tomorrow morning.

**Comment by obfuscurity** [comment]
Just a general note that this is a hard blocker for the next release.

**Comment by deniszh** [comment]
I have plans to use newly-opensourced pyflame profiler http://eng.uber.com/pyflame/ to find where culprit is.

**Comment by obfuscurity** [comment]
I've been looking at this for a while and I'm not able to reproduce anything that suggests a threading performance or deadlock issue. @Kenterfie Is it possible that the pickle rendering (or network transit) is taking longer than expected?

Here are some sample datapoints requesting one large query at a time. There's a single frontend with three backend storage nodes.

```
C0 - laptop client
B0 - backend storage
B1 - backend storage
B2 - backend storage
F0 - frontend render
```
```
(C0) $ curl -k "https://${F0}/render/?target=collectd.*.cpu-*.cpu-*&format=json" 2>&1 >/dev/null

(B0) Wed Oct 05 15:12:25 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 0.364455
(B0) Wed Oct 05 15:12:25 2016 :: Total pickle rendering time 0.424098
(B1) Wed Oct 05 15:12:25 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 0.556552
(B1) Wed Oct 05 15:12:25 2016 :: Total pickle rendering time 0.634059
(B2) Wed Oct 05 15:12:23 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 0.185500
(B2) Wed Oct 05 15:12:23 2016 :: Total pickle rendering time 0.206219
(F0) Wed Oct 05 15:12:25 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 1.124653
(F0) Wed Oct 05 15:12:27 2016 :: Total json rendering time 3.208207

(B0) Wed Oct 05 15:14:24 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 0.334817
(B0) Wed Oct 05 15:14:24 2016 :: Total pickle rendering time 0.392128
(B1) Wed Oct 05 15:14:24 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 0.467109
(B1) Wed Oct 05 15:14:24 2016 :: Total pickle rendering time 0.541729
(B2) Wed Oct 05 15:14:22 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 0.166958
(B2) Wed Oct 05 15:14:22 2016 :: Total pickle rendering time 0.186709
(F0) Wed Oct 05 15:14:24 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 0.802444
(F0) Wed Oct 05 15:14:26 2016 :: Total json rendering time 2.822548

(B0) Wed Oct 05 15:50:51 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 0.307690
(B0) Wed Oct 05 15:50:51 2016 :: Total pickle rendering time 0.366291
(B1) Wed Oct 05 15:50:51 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 0.498106
(B1) Wed Oct 05 15:50:51 2016 :: Total pickle rendering time 0.573213
(B2) Wed Oct 05 15:50:49 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 0.170795
(B2) Wed Oct 05 15:50:49 2016 :: Total pickle rendering time 0.190676
(F0) Wed Oct 05 15:50:51 2016 :: Retrieval of collectd.*.cpu-*.cpu-* took 0.888467
(F0) Wed Oct 05 15:50:53 2016 :: Total json rendering time 2.889416
```

Note that I believe that **something** is causing a performance regression for you guys, but it's been very difficult to reproduce the conditions in such a way that will point to *precisely where* we should begin looking.

**Comment by obfuscurity** [comment]
Apologies, I misunderstood the problem. I thought we were trying to track down a gap in time between the _backend retrieval_ finished and the _rendering_ began. I know see that we're talking about how long retrieval takes on the frontend node for local metrics _when clustering is enabled_.

**Comment by obfuscurity** [comment]
Regardless, I'm still not seeing any "regression" between master and 0.9.15. Both systems behave identically for the tests I've performed.

**Comment by Kenterfie** [comment]
Thanks @obfuscurity for your tests. This shows me the exact same behavior like i noticed. The regression is very low, but the problem itself exist. The time differences between backend and frontend are in my opinion to high. In my case mostly factor 10. I dont know if you want to fix this in the next version, but currently is it impossible to use the graphite cluster like intended. I have to use the backends directly to bypass the performance issue.

**Comment by obfuscurity** [comment]
@Kenterfie My results do **not** correlate to your findings. The `Retrieval` timer reported by each *frontend* represents the total (or at least, the time required to merge the required union) of all `CLUSTER_SERVERS` (and local) retrievals. Hence, it's only natural that disabling `CLUSTER_SERVERS` on the frontend would result in drastically reduced query times (it's only retrieving locally).

This is what I find so frustrating... I'm unable to reproduce anything that resembles your or @reyjrar's results.

**Comment by reyjrar** [comment]
Further testing based on slack call with @obfuscurity : 

https://gist.github.com/reyjrar/4c0ad075a528d4d68e14b8b1427a99be

Setup is 2 local nodes in each DC, 2 DC's.

**Comment by reyjrar** [comment]
Testing the effect of CLUSTER_SERVERS recursively

`# Front with CLUSTER_SERVERS, Backend w/o
Thu Oct 06 14:39:07 2016 :: Retrieval of carbon.agents.*.updateOperations took 0.116185
Thu Oct 06 14:39:07 2016 :: Rendered PNG in 0.275612 seconds
Thu Oct 06 14:39:07 2016 :: Total rendering time 0.393449 seconds
`

`# Front and Back ends with CLUSTER_SERVERS set to eachother's IPs
Thu Oct 06 14:37:25 2016 :: Retrieval of carbon.agents.*.updateOperations took 0.059602
Thu Oct 06 14:37:26 2016 :: Rendered PNG in 0.266713 seconds
Thu Oct 06 14:37:26 2016 :: Total rendering time 0.327662 seconds
`

These results weren't statistically significant.

**Comment by reyjrar** [comment]
I'm wondering if there's an optimization to be done, ie, keep-alive or connection pooling?  I can only venture a guess that the whole TCP setup/teardown to the CLUSTER_SERVERS is happening **before** the results are returned and/or merged?

**Comment by reyjrar** [comment]
So, I noticed this performance hit is real in `/metrics/find` as well.  So I diverted my resources there.  I think, and I might be wrong, that the caching in remote_storage.py is broken.  I am *never* seeing a cached result.

I don't have MEMCACHED_HOSTS set, so I noticed that my CACHES dict is empty.  I am attempting to override that local_settings.py with:

`
CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.db.DatabaseCache',
        'LOCATION': 'caching',
    }
}
`
Then I ran:
`django-admin.py --pythopath=<WEBAPP> --settings=graphite.settings createcachetable caching`

This created the caching table in my database.  But after several minutes of making requests, I'm not seeing anything in the caching table of my database.

**Comment by obfuscurity** [comment]
It would appear that the remote fetches are **not** happening in parallel, but sequentially. Note the results in the gist below: the first few (one for each of my `CLUSTER_SERVERS`) represent the initial connection setup; subsequent fetches reuse these connections and are much faster, but all of them are still blocking in order.

@Kenterfie would you mind applying the `datalib.py` patch here and gisting your results?

https://gist.github.com/obfuscurity/9f5f6c59d700a6ec1ec41585bc8ecaf8

**Comment by cbowman0** [comment]
I don't think that is how it works, @obfuscurity.  My understanding is that the loop starting with `for node, results in fetches:` is serializing the consumption of the results of the fetch requests.  Those will always happen in sequential order and the first one (or few) *should* take longer than the rest, since they are going to block on the remote store responding.  To calculate how long each remote fetch takes (along with the start/stop times), the instrumentation would need to be added to `node.fetch()` or within the `fetches = [(node, node.fetch(startTime, endTime)) for node in matching_nodes if node.is_leaf]` line as that's when the start of the remote request happens.

The `datalib.py` code in master is rather different from the one in the 0.9.x tree, but the major parts I see are:
1. In 0.9.x, STORE.find is only called for the local store.  All of the remote nodes are sent the fetch request without sending a find() first.  In master, we reduce the remote stores that we issue the fetch request to by first sending a find request in the line that assigns matching_nodes.
2. In 0.9.x the carbonlink supports a bulk query option, but master does not.
3. 0.9.x has as prefetch of remote data, but master does not.

With regards to # 1, I think that's a wash because the subsequent fetch() will result in a local find() call anyways on the remote nodes and that one should be cached and/or faster because the data would be in kernel's disk cache.  We could test it bypassing those find calls by converting STORE.remote_stores into the datatype that matching_nodes list contains.

For # 2, I'm not sure where in the master code to reimplement that feature.  Maybe in datalib.py:fetch_data() we can loop over matching_nodes and extract all metrics found, do the bulk carbonlink and pass that cached data to every WhisperReader.fetch(), as an example? 

For # 3, I don't know enough about the implementation of that to know if it would make a difference here.

**Comment by obfuscurity** [comment]
@cbowman0 I absolutely agree that the per-node fetches are serialized. However, if you look at my logs, it would appear that **all** of the fetches are serial, according to the timestamps (one does not start until the previous is finished). I have a hard time believing that none of these fetches would have intertwined run times. Please correct me if you think I'm misreading the output.

**Comment by obfuscurity** [comment]
Also, agreed on the http connection startup time. I wasn't implying that anything is wrong there, just highlighting it.

**Comment by obfuscurity** [comment]
@Kenterfie Were you running 0.9.x with `REMOTE_PREFETCH_DATA` enabled or not?

**Comment by obfuscurity** [comment]
@Kenterfie ping ^

**Comment by Kenterfie** [comment]
Sorry was busy. I have applied the patch above and get these outputs.

Thu Nov 17 17:45:02 2016 :: DEBUG: started all fetches at 1479401102.64
Thu Nov 17 17:45:02 2016 :: DEBUG: started all fetches at 1479401102.64
Thu Nov 17 17:45:02 2016 :: DEBUG: completed fetch in  0.092004 at 1479401102.7
Thu Nov 17 17:45:02 2016 :: DEBUG: started fetch at 1479401102.7
Thu Nov 17 17:45:02 2016 :: DEBUG: completed fetch in  0.000179 at 1479401102.7
Thu Nov 17 17:45:02 2016 :: DEBUG: started fetch at 1479401102.7
Thu Nov 17 17:45:02 2016 :: DEBUG: completed fetch in  0.000188 at 1479401102.7
Thu Nov 17 17:45:02 2016 :: DEBUG: started fetch at 1479401102.7
Thu Nov 17 17:45:02 2016 :: DEBUG: completed fetch in  0.000193 at 1479401102.7
Thu Nov 17 17:45:02 2016 :: DEBUG: started fetch at 1479401102.7
Thu Nov 17 17:45:02 2016 :: DEBUG: completed fetch in  0.000211 at 1479401102.7
Thu Nov 17 17:45:02 2016 :: DEBUG: completed all fetches in  1.048134 at 1479401102.7
Thu Nov 17 17:45:02 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.xxxx.yyyy.tt.*),4) took 1.062054
Thu Nov 17 17:45:02 2016 :: DEBUG: started all fetches at 1479401102.72
Thu Nov 17 17:45:02 2016 :: DEBUG: started fetch at 1479401102.97
Thu Nov 17 17:45:03 2016 :: DEBUG: completed fetch in  0.091993 at 1479401103.07
Thu Nov 17 17:45:03 2016 :: DEBUG: started fetch at 1479401103.07
Thu Nov 17 17:45:03 2016 :: DEBUG: completed fetch in  0.000132 at 1479401103.07
Thu Nov 17 17:45:03 2016 :: DEBUG: started fetch at 1479401103.07
Thu Nov 17 17:45:03 2016 :: DEBUG: completed fetch in  0.000252 at 1479401103.07
Thu Nov 17 17:45:03 2016 :: DEBUG: started fetch at 1479401103.07
Thu Nov 17 17:45:03 2016 :: DEBUG: completed fetch in  0.000168 at 1479401103.07
Thu Nov 17 17:45:03 2016 :: DEBUG: started fetch at 1479401103.07
Thu Nov 17 17:45:03 2016 :: DEBUG: completed fetch in  0.000159 at 1479401103.07
Thu Nov 17 17:45:03 2016 :: DEBUG: started fetch at 1479401103.07
Thu Nov 17 17:45:03 2016 :: DEBUG: completed fetch in  0.000157 at 1479401103.07
Thu Nov 17 17:45:03 2016 :: DEBUG: completed all fetches in  1.045338 at 1479401103.07
Thu Nov 17 17:45:03 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.xxxx.yyyy.tt2.*),4) took 1.054909
Thu Nov 17 17:45:03 2016 :: DEBUG: started all fetches at 1479401103.08
Thu Nov 17 17:45:03 2016 :: DEBUG: started fetch at 1479401103.33
Thu Nov 17 17:45:03 2016 :: DEBUG: started fetch at 1479401103.33
Thu Nov 17 17:45:03 2016 :: DEBUG: started fetch at 1479401103.39
Thu Nov 17 17:45:03 2016 :: DEBUG: completed fetch in  0.088522 at 1479401103.42
Thu Nov 17 17:45:03 2016 :: DEBUG: started fetch at 1479401103.42
Thu Nov 17 17:45:03 2016 :: DEBUG: completed fetch in  0.000198 at 1479401103.42
Thu Nov 17 17:45:03 2016 :: DEBUG: started fetch at 1479401103.42

@obfuscurity REMOTE_PREFETCH_DATA is disabled in my case

**Comment by Kenterfie** [comment]
Thu Nov 17 17:58:17 2016 :: DEBUG: started all fetches at 1479401897.07
Thu Nov 17 17:58:17 2016 :: DEBUG: started fetch at 1479401897.14
Thu Nov 17 17:58:17 2016 :: DEBUG: started fetch at 1479401897.15
Thu Nov 17 17:58:17 2016 :: DEBUG: completed fetch in  0.091928 at 1479401897.24
Thu Nov 17 17:58:17 2016 :: DEBUG: started fetch at 1479401897.24
Thu Nov 17 17:58:17 2016 :: DEBUG: completed fetch in  0.000256 at 1479401897.24
Thu Nov 17 17:58:17 2016 :: DEBUG: started fetch at 1479401897.24
Thu Nov 17 17:58:17 2016 :: DEBUG: completed fetch in  0.000179 at 1479401897.24
Thu Nov 17 17:58:17 2016 :: DEBUG: started fetch at 1479401897.24
Thu Nov 17 17:58:17 2016 :: DEBUG: completed fetch in  0.000206 at 1479401897.24
Thu Nov 17 17:58:17 2016 :: DEBUG: started fetch at 1479401897.24
Thu Nov 17 17:58:17 2016 :: DEBUG: completed fetch in  0.000199 at 1479401897.24
Thu Nov 17 17:58:17 2016 :: DEBUG: completed all fetches in  1.041527 at 1479401897.24

Why are so many fetches for one query? This makes no sense to me, what cause the hugh delay of >1 second. Only two web api endpoints are defined, so only to fetches.

**Comment by obfuscurity** [comment]
I've looked at this for the better part of 3 weeks now, attempting to forward port the [parallel fetch](https://github.com/graphite-project/graphite-web/pull/1026) and [prefetch](https://github.com/graphite-project/graphite-web/pull/1073) designs from `0.9.x` to `master`. I'm not embarrassed to admit that I'm not smart enough to figure it out on my own.

I've been pushing very hard to release 1.0.0, but this behavioral regression is simply too important to ignore. For the time being I'm planning to bump [setup.py](https://github.com/graphite-project/graphite-web/blob/master/setup.py#L72) to 0.10.0-rc2 and ask folks to test while we continue to look into this cluster performance issue.

**Comment by obfuscurity** [comment]
The wonderful @DanCech has been kind enough to hack on this for a while. He has a branch up that seems to be fixing the clustering performance issue. @Kenterfie @reyjrar @redbaron would you be so kind as to try out his branch and see if that improves things for you?

https://github.com/graphite-project/graphite-web/compare/master...DanCech:clustering

@graphite-project/committers This is going to be a significant patch so it would be great to get eyes on it sooner rather than later.

**Comment by bmhatfield** [comment]
Can we do this without pulling in `requests`? We've been pretty good about avoiding it thus far as a requirement. Or we should go all-in on it.

**Comment by obfuscurity** [comment]
Per conversations with @DanCech, it seems to be the only http lib that's thread-safe.

**Comment by DanCech** [comment]
I'm going to see if I can use urllib3 directly, since that's what requests uses anyway.  httplib and urllib2 aren't thread safe.

**Comment by DanCech** [comment]
bear in mind the current patch is very much a work in progress, I do intend to migrate the http calls for delegated rendering etc to the same lib

**Comment by Kenterfie** [comment]
I have tried the branch above, but without noticable performance improvement in my case. The timings are nearly the same. I would help more, when i had the time to write a own implementation currently, but thanks for your effort to fix this bottleneck. In my opinion the http fetch himself is not the reason of the hugh render times on frontend side.

Less than 10ms on backend (rendering) + ~60ms transmit of data + magic on webapp frontend (800ms)

**Comment by DanCech** [comment]
@Kenterfie can you post the log output for an example request to the render endpoint on your system?

**Comment by Kenterfie** [comment]
Tue Nov 29 11:22:17 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.YXZ.*),4) took 1.055472
Tue Nov 29 11:22:17 2016 :: Total json rendering time 1.059517
Tue Nov 29 11:22:20 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.YXG.*),4) took 1.035930
Tue Nov 29 11:22:20 2016 :: Total json rendering time 1.039196
Tue Nov 29 11:22:22 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.YRG.*),4) took 1.079519
Tue Nov 29 11:22:22 2016 :: Total json rendering time 1.083786
Tue Nov 29 11:22:23 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.EFG.*),4) took 1.045477
Tue Nov 29 11:22:23 2016 :: Total json rendering time 1.053923
Tue Nov 29 11:22:28 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.EEG.*),4) took 1.054044
Tue Nov 29 11:22:28 2016 :: Total json rendering time 1.058418
Tue Nov 29 11:22:30 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.EED.*),4) took 1.051837
Tue Nov 29 11:22:30 2016 :: Total json rendering time 1.058412
Tue Nov 29 11:22:32 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.EDS.*),4) took 1.053786
Tue Nov 29 11:22:32 2016 :: Total json rendering time 1.058767
Tue Nov 29 11:22:34 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.XDF.*),4) took 1.065419
Tue Nov 29 11:22:34 2016 :: Total json rendering time 1.069655
Tue Nov 29 11:22:35 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.RGH.*),4) took 1.050324
Tue Nov 29 11:22:35 2016 :: Total json rendering time 1.054232
Tue Nov 29 11:22:39 2016 :: Retrieval of aliasByNode(keepLastValue(statsite.NBG.*),4) took 0.880803
Tue Nov 29 11:22:39 2016 :: Total json rendering time 0.884234

all renderings around the 1 second on frontend server

Tue Nov 29 10:51:33 2016 :: Retrieval of statsite.NBG.* took 0.010941
Tue Nov 29 10:51:33 2016 :: Total pickle rendering time 0.012926

~10ms on backend

other logs later, but you can see, they dont very different from logs before

**Comment by DanCech** [comment]
@Kenterfie I just pushed some updates, can you try again with `DEBUG = True` and `LOG_RENDERING_PERFORMANCE = True` and post the output of `tail -qF info.log exception.log rendering.log` for a single request?

**Comment by Kenterfie** [comment]
sure, i do this later after my main work

**Comment by Kenterfie** [comment]
https://gist.github.com/Kenterfie/32b3041bc29bea87aa94d0c826b149e0 here are anonymised log with your changes @DanCech

**Comment by DanCech** [comment]
can you grab the latest code? we identified a couple of issues today.  Also can you make the backend IPs distinguishable? 0.0.0.0 0.0.0.1 etc vs all 0.0.0.0 would be great.

**Comment by Kenterfie** [comment]
is your latest code and i have updated now the log from above

**Comment by DanCech** [comment]
@Kenterfie are you sure? the log lines like:
```
Tue Nov 29 22:53:37 2016 :: FindRequest.send(host=0.0.158.22:1444, query=<FindQuery: statsite.anonym.data.* from Tue Nov 29 21:53:37 2016 until Tue Nov 29 22:53:37 2016>) called
```
should include a raw timestamp at the end like:
```
Tue Nov 29 17:39:01 2016 :: FindRequest.send(host=1.2.3.4:80, query=<FindQuery: carbon.agents.*.* from Wed Nov 23 10:12:46 2016 until Thu Nov 24 10:12:46 2016>) called at 1480459141.4
```

**Comment by Kenterfie** [comment]
Sorry was my mistake. Was on your master branch. Now i have your clustering branch. Here you can find log.

https://gist.github.com/Kenterfie/52c2f8caacdcbb9a06c3f86348a2c15f

**Comment by OrangeDog** [comment]
If #1768 is merged, this is fixed now, right?

**Comment by deniszh** [comment]
@OrangeDog - no, it's not connected with #1768

**Comment by iksaif** [comment]
Could anybody affect by this issue try the latest HEAD and give us feedback? Thanks !

**Comment by deniszh** [comment]
Graphite 0.9.16 and 1.0.0 are released. Please report any performance issue (and other issues too, of course)

**Comment by stan-sack** [comment]
@iksaif @deniszh this is still definitely an issue. I just spent around 3 weeks digging through render logs and trying different version/cluster configurations after upgrading a 0.9.15 installation to 1.1.2, and finding a relatively large expression with 2 wildcards (approx 1000 metrics) began to time out. 

I set up 3 different graphite-web instances fetching from the same 3 node carbonlink cluster over http. The expression would render in around 50 seconds on 0.9.15, 220 seconds on 1.0.2, and take over 600 (timeout here) seconds on 1.1.2. This was on cold start with no cache and version 0.9.15 running on the carbonlink cluster nodes.

**Comment by DanCech** [comment]
@stan-sack if you can share some logs that would help us figure out what's happening in your case

**Comment by reyjrar** [comment]
Would also like to confirm, I'm updated to 1.1.2 and recently moved to using clustering over HTTP and have the same issue.  1.x and onward is *dramatic* performance degradation.  I even tried adding memcached caches in to relieve some of the slowness, but that only works for queries in the caches.

I'll post logs soon from the render on the graphite-web nodes talking directly to carbon-cache's over carbonlink and of the "front end" nodes I have communicating to the backend servers via CLUSTER_SERVERS.

**Comment by deniszh** [comment]
@reyjrar @stan-sack :
Could you please open a new issue for that, to increase its visibility? It's probably not related to original (and closed) issue anyway.
Thanks a lot!
